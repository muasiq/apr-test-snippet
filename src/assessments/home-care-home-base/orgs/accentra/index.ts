import { ConfigurationType } from '@prisma/client';
import { PathwaysNestingItem } from '../../../../server/api/routers/patient/v1/custom-form-responses/careplan';
import { Configuration } from '../../../configurations';
import { Pathway } from '../../data-files/scripts/clinicalPathways';
import { accentraPathwaysNesting } from './assets/sn-nesting';
import { backOfficeCompletedConfig } from './backOfficeCompletedConfig';
import { allQuestions } from './data-files';
import pathways from './data-files/clinicalPathways.json';
import { qaConfig } from './qaConfig';

export const AccentraConfig: Configuration = {
	orgName: 'Accentra',
	configurationType: ConfigurationType.HomeCareHomeBase,
	qaConfig,
	backOfficeCompletedConfig,
	allQuestions,
	pathways: pathways as Pathway[],
	pathwaysNesting: accentraPathwaysNesting as PathwaysNestingItem[],
};
