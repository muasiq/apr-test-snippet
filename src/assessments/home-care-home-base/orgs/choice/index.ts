import { ConfigurationType } from '@prisma/client';
import { PathwaysNestingItem } from '../../../../server/api/routers/patient/v1/custom-form-responses/careplan';
import { Configuration } from '../../../configurations';
import { choicePathwaysNesting } from './assets/sn-nesting';
import { backOfficeCompletedConfig } from './backOfficeCompletedConfig';
import { allQuestions } from './data-files';
import pathways from './data-files/clinicalPathways.json';
import { qaConfig } from './qaConfig';

export const ChoiceConfig: Configuration = {
	orgName: 'Choice',
	configurationType: ConfigurationType.Choice,
	qaConfig,
	backOfficeCompletedConfig,
	allQuestions,
	pathways,
	pathwaysNesting: choicePathwaysNesting as PathwaysNestingItem[],
};
