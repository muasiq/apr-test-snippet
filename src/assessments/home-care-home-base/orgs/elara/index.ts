import { ConfigurationType } from '@prisma/client';
import { PathwaysNestingItem } from '../../../../server/api/routers/patient/v1/custom-form-responses/careplan';
import { Configuration } from '../../../configurations';
import { Pathway } from '../../data-files/scripts/clinicalPathways';
import { elaraPathwaysNesting } from './assets/sn-nesting';
import { backOfficeCompletedConfig } from './backOfficeCompletedConfig';
import { allQuestions } from './data-files';
import pathways from './data-files/clinicalPathways.json';
import { qaConfig } from './qaConfig';

export const ElaraConfig: Configuration = {
	orgName: 'Elara',
	configurationType: ConfigurationType.HomeCareHomeBase,
	qaConfig,
	backOfficeCompletedConfig,
	allQuestions,
	pathways: pathways as Pathway[],
	pathwaysNesting: elaraPathwaysNesting as PathwaysNestingItem[],
};
