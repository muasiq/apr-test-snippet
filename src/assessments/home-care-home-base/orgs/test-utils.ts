import { Group, SubGroup } from '~/assessments/types';

export function allIdsInConfig(group: Group[]) {
	const allQuestions = allQuestionsInConfig(group);
	return allQuestions.map((q) => q.id);
}

export function allQuestionsInConfig(group: Group[]) {
	return group.flatMap((g) => {
		if (g.subHeadings) {
			return g.subHeadings.flatMap((sh) => sh.subGroups.flatMap((sg) => allSubGroupQuestions(sg)));
		}
		return g.subGroups.flatMap((sg) => allSubGroupQuestions(sg));
	});
}

export function allSubGroupQuestions(sg: SubGroup) {
	if (sg.multiple) {
		return sg.multiple.questions.filter((q) => q.id);
	}
	return sg.questions.filter((q) => q.id);
}
