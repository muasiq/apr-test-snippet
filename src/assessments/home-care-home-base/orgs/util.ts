/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unsafe-argument */
import * as fs from 'fs';
import { compact } from 'lodash';
import { getQuestionBank } from '~/assessments/apricot/scripts/uploadNewItemsToAllQuestionsBank';
import { listDocs } from '../../scripts/docs';
import { getSheetData } from '../../scripts/sheets';
import { parseAnswersFile, toReplace } from '../data-files/scripts/createMasterQuestionMap';

export async function makeFile(config: string, tab: string, orgName: string, driveFolderId?: string) {
	const data = await getSheetData(config, tab);
	const toWrite = await Promise.all(
		data.rows.map(async (row) => {
			const data = row.toObject();
			if (data.responses) {
				if (tab === 'soc-standard-assessment' && isLargerResponse(data) && driveFolderId) {
					await handleDownloadingLargerResponses(data, 'Allergies', driveFolderId);
				}
				try {
					const parsed = JSON.parse(data.responses) as any[];
					parsed.map((p: any) => {
						if (p.followUp) {
							p.followup = p.followUp;
							delete p.followUp;
						}
						if (p.followup?.length && !Array.isArray(p.followup) && !p.followup.startsWith('[')) {
							p.followup = [p.followup];
						}
						if (!p.followup) {
							delete p.followup;
						} else {
							p.followup = compact(p.followup?.map((f: string) => f.trim()));
							if (tab === 'physical-assessment') {
								const oasisFollowUpsNotNeeded = [
									'PA-d50b20',
									'PA-ed3aea',
									'PA-6b551a',
									'PA-d1f1f6',
									'PA-0b3c1f',
									'PA-b8f0e4',
								];
								p.followup = p.followup.filter((f: string) => !oasisFollowUpsNotNeeded.includes(f));
							}
						}
					});
					return { ...data, responses: parsed };
				} catch (e) {
					console.log(data);
					console.error(e);
					throw e;
				}
			} else {
				delete data.responses;
			}
			return data;
		}),
	);
	if (tab === 'soc-standard-assessment') {
		const allergiesQuestion = toWrite.find((question) => question.id === 'MQ67LVO5CK');
		if (!allergiesQuestion) throw new Error('Allergies question not found');
		const otherResponse = allergiesQuestion.responses.find((response: any) => response.text === 'Other');
		if (!otherResponse) {
			allergiesQuestion.responses.push({
				text: 'Other',
				followup: ['MQ3ZFBFY4B'],
			});
		}
	}
	if (tab === 'physical-assessment' || tab === 'soc-standard-assessment') {
		await updateCleanedTextFromBank(toWrite, tab);
	}
	fs.writeFileSync(
		`src/assessments/home-care-home-base/orgs/${orgName}/data-files/${tab}.json`,
		JSON.stringify(toWrite, null, 2),
	);
}

async function updateCleanedTextFromBank(data: Partial<Record<string, any>>[], tab: string) {
	const bankItems = await getQuestionBank();
	data.forEach((d) => {
		const bankItem = bankItems.find((b) => b.id === d.id);
		if (bankItem) {
			if (tab === 'physical-assessment') {
				d.cleanedText = bankItem.cleanedText?.trim();
			} else if (tab === 'soc-standard-assessment') {
				d.originalText = bankItem.originalText?.trim();
			}
		}
	});
}

function isLargerResponse(data: Partial<Record<string, any>>) {
	const text = data.text;
	const matchesLookup = toReplace[text];
	if (matchesLookup && matchesLookup === 'Allergies') {
		return true;
	}
	return false;
}

async function handleDownloadingLargerResponses(
	data: Partial<Record<string, any>>,
	replaceMatch: string,
	driveFolderId: string,
) {
	const allFiles = await listDocs(driveFolderId);
	if (!allFiles) throw new Error('No files found in folder');
	const newResponses = await parseAnswersFile(replaceMatch, allFiles);
	data.responses = JSON.stringify(newResponses);
}
