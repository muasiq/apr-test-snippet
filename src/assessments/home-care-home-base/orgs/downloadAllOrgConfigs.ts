import { downloadAccentraConfiguration } from './accentra/downloadAccentraConfiguration';
import { downloadChoiceConfiguration } from './choice/downloadChoiceConfiguration';
import { downloadElaraConfiguration } from './elara/downloadElaraConfiguration';

async function main() {
	await downloadAccentraConfiguration();
	await downloadChoiceConfiguration();
	await downloadElaraConfiguration();
}

main().catch(console.error);
