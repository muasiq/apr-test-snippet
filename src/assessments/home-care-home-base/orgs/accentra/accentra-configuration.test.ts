import { expect, it } from '@jest/globals';
import { allIdsInConfig, allQuestionsInConfig } from '~/assessments/home-care-home-base/orgs/test-utils';
import { recursivelyGetAllQuestionsAndFollowUps } from '~/assessments/util/lookupQuestionUtil';
import { backOfficeCompletedConfig } from './backOfficeCompletedConfig';
import { AccentraConfig } from './index';
import { qaConfig } from './qaConfig';

const allQuestions = AccentraConfig.allQuestions.filter((q) => {
	if (q.type) {
		return q.type.toLowerCase() !== 'panel';
	}
	return true;
});

describe('Ensure Accentra Configuration is Valid', () => {
	it('Should not have questions without ids', () => {
		allQuestions.forEach((q) => {
			expect(q.id).toBeDefined();
			expect(q.id.length).toBeGreaterThan(0);
		});
	});

	it('Should not have ids in one config and not the other', () => {
		const allQaConfigIds = allIdsInConfig(qaConfig);
		const allCompletedConfigIds = allIdsInConfig(backOfficeCompletedConfig);
		const idsInQANotInCompleted = allQaConfigIds.filter((id) => !allCompletedConfigIds.includes(id));
		const idsInCompletedNotInQA = allCompletedConfigIds.filter((id) => !allQaConfigIds.includes(id));
		// console.log({ idsInQANotInCompleted, idsInCompletedNotInQA });
		expect(idsInQANotInCompleted).toEqual([]);
		expect(idsInCompletedNotInQA).toEqual([]);
	});

	it('All ids in config (and follow-ups) should be in datafile', () => {
		const allQaConfigQuestions = allQuestionsInConfig(qaConfig);
		const allQaConfigQuestionsAndFollowups = recursivelyGetAllQuestionsAndFollowUps(
			allQaConfigQuestions,
			AccentraConfig,
		);
		for (const id of allQaConfigQuestionsAndFollowups.map((q) => q.id)) {
			const match = AccentraConfig.allQuestions.find((q) => q.id === id);
			if (!match) {
				console.log(`missing id: ${id}`);
			}
			expect(match).toBeDefined();
		}
	});

	it('Should not have any top level questions in data file not in config', () => {
		const allFollowUps = allQuestions.flatMap((q) => q.responses?.flatMap((r) => r.followup) ?? []);
		const topLevel = allQuestions.filter((q) => !allFollowUps.includes(q.id));
		const allQaConfigIds = allIdsInConfig(qaConfig);
		const missingFromConfig = topLevel.filter((q) => !allQaConfigIds.includes(q.id));
		const missingFromConfigIds = missingFromConfig.map((q) => q.id);
		expect(missingFromConfigIds.length).toEqual(78);
		expect(missingFromConfigIds).toMatchSnapshot();
	});

	it('Should not have follow ups in the config', () => {
		const allQaConfigIds = allIdsInConfig(qaConfig);
		const allFollowUps = allQuestions
			.flatMap((q) => q.responses?.flatMap((r) => r.followup) ?? [])
			.filter((f) => f) as string[];
		const followUpsInConfig = allFollowUps.filter((id) => allQaConfigIds.includes(id));
		expect(followUpsInConfig).toEqual([]);
	});

	it('Should not have duplicate ids in the qaConfig', () => {
		const allQaConfigIds = allIdsInConfig(qaConfig);
		const duplicates = allQaConfigIds.filter((id, index) => allQaConfigIds.indexOf(id) !== index);
		expect(duplicates).toEqual([]);
	});

	it('Completed Config Duplicate Report', () => {
		const allQaConfigIds = allIdsInConfig(backOfficeCompletedConfig).filter((id) => id) as string[];
		const list: { id: string; numberOfTimesDuplicated: number }[] = [];
		allQaConfigIds.forEach((id) => {
			const numberOfTimesDuplicated = allQaConfigIds.filter((i) => i === id).length;
			if (!list.find((i) => i.id === id)) {
				list.push({ id, numberOfTimesDuplicated });
			}
		});
		// const ordered = list.sort((a, b) => b.numberOfTimesDuplicated - a.numberOfTimesDuplicated);
		// fs.writeFileSync('temp.json', JSON.stringify(ordered, null, 2));
	});

	it('Should merge in any oasis followups', () => {
		const questionWithMergedFollowup = allQuestions.find((q) => q.id === 'M1340');
		const responseWithFollowup = questionWithMergedFollowup?.responses?.find(
			(r) => r.text === 'Yes, patient has at least one observable surgical wound',
		);
		if (!responseWithFollowup?.followup) {
			throw new Error('failed test - responseWithFollowup not found');
		}
		expect(responseWithFollowup.followup.length).toBe(2);
		expect(responseWithFollowup.followup[1]).toBe('PA-96dadc');
	});
});
