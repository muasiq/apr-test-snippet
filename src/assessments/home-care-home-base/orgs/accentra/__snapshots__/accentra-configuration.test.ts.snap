// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ensure Accentra Configuration is Valid Should not have any top level questions in data file not in config 1`] = `
[
  "AC1000",
  "AC1039",
  "AC1046",
  "AC1059E",
  "AC1069E",
  "AC1079E",
  "AC1090E",
  "AC1",
  "M0018",
  "M0010",
  "M0014",
  "M0016",
  "M0020",
  "M0040_A",
  "M0040_B",
  "M0040_C",
  "M0040_D",
  "M0050",
  "M0060",
  "M0064",
  "M0063",
  "M0065",
  "M0069",
  "M0066",
  "M0150",
  "M0150_specify",
  "A1110A",
  "M0030",
  "M0080",
  "M0090",
  "M0100",
  "M0102",
  "M0104",
  "M0110",
  "C0500",
  "D0160",
  "M1021_A2_ICD",
  "M1021_A2_SEVERITY",
  "M1023_B2_ICD",
  "M1023_B2_SEV",
  "M1023_C2_ICD",
  "M1023_C2_SEV",
  "M1023_D2_ICD",
  "M1023_D2_SEV",
  "M1023_E2_ICD",
  "M1023_E2_SEV",
  "M1023_F2_ICD",
  "M1023_F2_SEV",
  "M1060A",
  "M1060B",
  "N0415_2",
  "O0110_A_old_copy",
  "PA-946407",
  "PA-ee1734",
  "PA-fa9e30",
  "PA-22f022",
  "PA-f83cf9",
  "PA-2c4dff",
  "PA-a5c924",
  "PA-a0c7d8",
  "PA-ec76c0",
  "PA-f7b8c7",
  "PA-35d4af",
  "PA-4cab6a",
  "PA-193d0e",
  "PA-f4f24b",
  "PA-6e5b44",
  "PA-6d3cd4",
  "PA-0efd08",
  "PA-e26b4b",
  "PA-f3cbc0",
  "PA-300edd",
  "PA-07ac47",
  "PA-383ffa",
  "PA-5e1b3a",
  "MQA80Z8BML",
  "MQ5RWEJETF",
  "med_admin_custom",
]
`;
